std/machines/hash/keccakf16_memory.asm

```diff
@@ -0,0 +1,876 @@
+    link if final_step ~> (addr_h[28], addr_l[28]) = add_sub.add(output_addr_h, output_addr_l, 0, 116);
+    link if final_step ~> (addr_h[29], addr_l[29]) = add_sub.add(output_addr_h, output_addr_l, 0, 120);
+    link if final_step ~> (addr_h[30], addr_l[30]) = add_sub.add(output_addr_h, output_addr_l, 0, 124);
+    link if final_step ~> (addr_h[31], addr_l[31]) = add_sub.add(output_addr_h, output_addr_l, 0, 128);
+    link if final_step ~> (addr_h[32], addr_l[32]) = add_sub.add(output_addr_h, output_addr_l, 0, 132);
+    link if final_step ~> (addr_h[33], addr_l[33]) = add_sub.add(output_addr_h, output_addr_l, 0, 136);
+    link if final_step ~> (addr_h[34], addr_l[34]) = add_sub.add(output_addr_h, output_addr_l, 0, 140);
+    link if final_step ~> (addr_h[35], addr_l[35]) = add_sub.add(output_addr_h, output_addr_l, 0, 144);
+    link if final_step ~> (addr_h[36], addr_l[36]) = add_sub.add(output_addr_h, output_addr_l, 0, 148);
+    link if final_step ~> (addr_h[37], addr_l[37]) = add_sub.add(output_addr_h, output_addr_l, 0, 152);
+    link if final_step ~> (addr_h[38], addr_l[38]) = add_sub.add(output_addr_h, output_addr_l, 0, 156);
+    link if final_step ~> (addr_h[39], addr_l[39]) = add_sub.add(output_addr_h, output_addr_l, 0, 160);
+    link if final_step ~> (addr_h[40], addr_l[40]) = add_sub.add(output_addr_h, output_addr_l, 0, 164);
+    link if final_step ~> (addr_h[41], addr_l[41]) = add_sub.add(output_addr_h, output_addr_l, 0, 168);
+    link if final_step ~> (addr_h[42], addr_l[42]) = add_sub.add(output_addr_h, output_addr_l, 0, 172);
+    link if final_step ~> (addr_h[43], addr_l[43]) = add_sub.add(output_addr_h, output_addr_l, 0, 176);
+    link if final_step ~> (addr_h[44], addr_l[44]) = add_sub.add(output_addr_h, output_addr_l, 0, 180);
+    link if final_step ~> (addr_h[45], addr_l[45]) = add_sub.add(output_addr_h, output_addr_l, 0, 184);
+    link if final_step ~> (addr_h[46], addr_l[46]) = add_sub.add(output_addr_h, output_addr_l, 0, 188);
+    link if final_step ~> (addr_h[47], addr_l[47]) = add_sub.add(output_addr_h, output_addr_l, 0, 192);
+    link if final_step ~> (addr_h[48], addr_l[48]) = add_sub.add(output_addr_h, output_addr_l, 0, 196);
+
+    // Expects input of 25 64-bit numbers decomposed to 25 chunks of 4 16-bit little endian limbs. 
+    // The output is a_prime_prime_prime_0_0_limbs for the first 4 and a_prime_prime for the rest.
+
+    // Write memory while converting output to big endian format.
+    // Specifically, output obtained from the keccak computation are little endian.
+    // However, this keccakf16_memory machine produces big endian outputs in memory.
+    // Therefore, memory write converts little endian from keccak computation to big endian for the output in memory.
+    link if final_step ~> mem.mstore(output_addr_h, output_addr_l, time_step, a_prime_prime_prime_0_0_limbs[3], a_prime_prime_prime_0_0_limbs[2]);
```

Comment:
@georgwiese how expensive will these be with logup? Can they be batched somehow?

Reply:
As @qwang98 pointed out in the text, the alternative would be to repeat the output in all rows of the block (needs 100 extra witness columns) and do up to 3 memory writes (`ceil(50 / 24)`) in each of the 24 rows of the block. The same holds of the `add_sub` links and the memory ready.

With all LogUp optimizations (which we have not implemented yet), the cost of a link would be `0.5 * <number of extension field elements>`, so 2 in the case of BabyBear. Also, each permutation currently adds a selector and new accumulation columns to the memory machine, but we could and should optimize that, so I think the extra cost in the memory machine would be zero.

So all in all not to bad, assuming a bunch of optimizations we should do anyway.

Reply:
I assume your mention of "each permutation currently adds a selector and new accumulation columns to the memory machine, but we could and should optimize that" refers to these lines? https://github.com/powdr-labs/powdr/blob/main/std/machines/small_field/memory.asm#L44-L45

How would we optimize this?